File: tools/src/import/listforge.ts
/**
 * ListForge adapter: lower a decoded ListForge "share JSON" payload (a
 * BattleScribe-derived roster tree) to a {@link ParsedRoster}.
 *
 * The walk reads an ALLOWLIST of fields only — `name`, `number`, `type`,
 * `categories[].name`, `group`, and `costs` point values — and never touches
 * `rules[].description` or ability `profiles[].characteristics[].$text`, which
 * carry reproduced rules text. This keeps the importer's output free of
 * copyrighted prose by construction.
 *
 * Selection-tree shape (recursive `selections`):
 * - Configuration nodes (`type: "upgrade"`) named "Detachment" / "Battle Size"
 *   carry the chosen value as their first child selection.
 * - Unit nodes (`type: "model" | "unit"`) carry role categories, a points cost,
 *   and — nested anywhere beneath them — their wargear (weapon-category
 *   selections), enhancement (a selection whose `group` starts "Enhancements"),
 *   the "Warlord" marker, and model sub-selections.
 * - Every unit carries a `"Faction: <Name>"` category.
 *
 * @packageDocumentation
 */
import type { FormatAdapter } from "./adapter.js";
import type { ParsedRoster, ParsedUnit, ParsedWargear } from "./types.js";

const PTS_COST_NAME = "pts";
const FACTION_CATEGORY = /^Faction:\s*(.+)$/;
const POINTS_LIMIT = /(\d[\d,]*)\s*Point/i;
const ENHANCEMENT_GROUP_PREFIX = "Enhancements";
const CHARACTER_CATEGORIES = new Set(["Character", "Epic Hero"]);
const WEAPON_CATEGORY_SUFFIX = " Weapon"; // "Ranged Weapon", "Melee Weapon", "Psychic Weapon"
const NEWRECRUIT_XMLNS = "http://www.battlescribe.net/schema/rosterSchema";
const NEWRECRUIT_HOST_PREFIX = "https://newrecruit";

// --- Minimal structural views of the parts of the payload we read. ----------

interface RawCategory {
  name?: unknown;
}
interface RawCost {
  name?: unknown;
  value?: unknown;
}
interface RawSelection {
  name?: unknown;
  type?: unknown;
  number?: unknown;
  group?: unknown;
  categories?: unknown;
  costs?: unknown;
  selections?: unknown;
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function asString(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function selectionName(sel: RawSelection): string {
  return asString(sel.name) ?? "";
}

function selectionType(sel: RawSelection): string {
  return asString(sel.type) ?? "";
}

/** A selection's multiplicity (`number`), defaulting to 1. */
function selectionCount(sel: RawSelection): number {
  return typeof sel.number === "number" && sel.number > 0 ? sel.number : 1;
}

/** Point value from a selection's cost block, or null when absent. */
function pointsOf(sel: RawSelection): number | null {
  for (const raw of asArray(sel.costs)) {
    const cost = raw as RawCost;
    if (asString(cost.name) === PTS_COST_NAME && typeof cost.value === "number") {
      return cost.value;
    }
  }
  return null;
}

function categoryNames(sel: RawSelection): string[] {
  return asArray(sel.categories)
    .map((c) => asString((c as RawCategory).name))
    .filter((n): n is string => n !== null);
}

function childSelections(sel: RawSelection): RawSelection[] {
  return asArray(sel.selections) as RawSelection[];
}

/** Depth-first visit of a selection and everything beneath it. */
function walk(sel: RawSelection, visit: (s: RawSelection) => void): void {
  visit(sel);
  for (const child of childSelections(sel)) walk(child, visit);
}

function isUnitSelection(sel: RawSelection): boolean {
  const type = selectionType(sel);
  return type === "model" || type === "unit";
}

function isCharacter(sel: RawSelection): boolean {
  return categoryNames(sel).some((n) => CHARACTER_CATEGORIES.has(n));
}

function isWeaponSelection(sel: RawSelection): boolean {
  return categoryNames(sel).some((n) => n.endsWith(WEAPON_CATEGORY_SUFFIX));
}

function isEnhancementSelection(sel: RawSelection): boolean {
  const group = asString(sel.group);
  return group !== null && group.startsWith(ENHANCEMENT_GROUP_PREFIX);
}

/** Sum the model count of a unit from its nested model selections. */
function modelCount(unit: RawSelection): number {
  let total = 0;
  walk(unit, (s) => {
    if (selectionType(s) === "model") total += selectionCount(s);
  });
  return total > 0 ? total : selectionCount(unit);
}

/** Build a parsed unit from a top-level unit selection. */
function parseUnit(unit: RawSelection): ParsedUnit {
  const wargear: ParsedWargear[] = [];
  let enhancement_raw_name: string | null = null;
  let enhancement_points: number | null = null;
  let is_warlord = false;

  for (const node of childSelections(unit)) {
    walk(node, (s) => {
      if (isEnhancementSelection(s)) {
        if (enhancement_raw_name === null) {
          enhancement_raw_name = selectionName(s);
          enhancement_points = pointsOf(s);
        }
        return;
      }
      if (selectionName(s) === "Warlord") {
        is_warlord = true;
        return;
      }
      if (isWeaponSelection(s)) {
        wargear.push({ raw_name: selectionName(s), count: selectionCount(s) });
      }
    });
  }

  return {
    raw_name: selectionName(unit),
    is_character: isCharacter(unit),
    model_count: modelCount(unit),
    points: pointsOf(unit),
    is_warlord,
    enhancement_raw_name,
    enhancement_points,
    wargear,
  };
}

/** Value carried as the first child of a named configuration selection. */
function configValue(
  selections: RawSelection[],
  configName: string,
): string | null {
  const node = selections.find((s) => selectionName(s) === configName);
  if (!node) return null;
  const child = childSelections(node)[0];
  return child ? selectionName(child) : null;
}

/** Every value under a named config, across repeated blocks and multiple
 * children, in source order. Used for multi-detachment 11e lists. */
function configValues(selections: RawSelection[], configName: string): string[] {
  const out: string[] = [];
  for (const node of selections) {
    if (selectionName(node) !== configName) continue;
    for (const child of childSelections(node)) {
      const name = selectionName(child);
      if (name) out.push(name);
    }
  }
  return out;
}

function parseLimit(label: string | null): number | null {
  if (!label) return null;
  const match = POINTS_LIMIT.exec(label);
  if (!match) return null;
  return Number.parseInt(match[1].replace(/,/g, ""), 10);
}

/** First `"Faction: X"` category found anywhere; reports all distinct names. */
function collectFactions(forces: RawSelection[]): string[] {
  const seen = new Set<string>();
  for (const force of forces) {
    for (const sel of childSelections(force)) {
      walk(sel, (s) => {
        for (const name of categoryNames(s)) {
          const match = FACTION_CATEGORY.exec(name);
          if (match) seen.add(match[1].trim());
        }
      });
    }
  }
  return [...seen];
}

interface RawRoster {
  name?: unknown;
  costs?: unknown;
  forces?: unknown;
  xmlns?: unknown;
  generatedBy?: unknown;
}
interface RawPayload {
  name?: unknown;
  generatedBy?: unknown;
  roster?: unknown;
}

function rosterOf(decoded: unknown): RawRoster | null {
  if (!decoded || typeof decoded !== "object") return null;
  const roster = (decoded as RawPayload).roster;
  if (!roster || typeof roster !== "object") return null;
  if (!Array.isArray((roster as RawRoster).forces)) return null;
  return roster as RawRoster;
}

/** Detect a NewRecruit-flavoured BattleScribe payload. ListForge's matcher
 * excludes these so the greedy first-match dispatcher routes them to the
 * NewRecruit adapter without falling through to here. */
function hasNewRecruitSignature(decoded: unknown, roster: RawRoster): boolean {
  if (asString(roster.xmlns) === NEWRECRUIT_XMLNS) return true;
  const genBy =
    asString((decoded as RawPayload).generatedBy) ?? asString(roster.generatedBy);
  return genBy !== null && genBy.toLowerCase().startsWith(NEWRECRUIT_HOST_PREFIX);
}

export const listForgeAdapter: FormatAdapter = {
  id: "listforge",

  matches(decoded: unknown): boolean {
    const roster = rosterOf(decoded);
    if (!roster) return false;
    return !hasNewRecruitSignature(decoded, roster);
  },

  parse(decoded: unknown): ParsedRoster {
    const payload = decoded as RawPayload;
    const roster = rosterOf(decoded);
    if (!roster) {
      throw new Error("listforge: payload has no roster.forces array");
    }

    const forces = asArray(roster.forces) as RawSelection[];

    // Configuration lives among each force's top-level selections.
    const detachment_raw_names: string[] = [];
    let battle_size_raw: string | null = null;
    const units: ParsedUnit[] = [];
    for (const force of forces) {
      const top = childSelections(force);
      detachment_raw_names.push(...configValues(top, "Detachment"));
      battle_size_raw ??= configValue(top, "Battle Size");
      for (const sel of top) {
        if (isUnitSelection(sel)) units.push(parseUnit(sel));
      }
    }

    const factions = collectFactions(forces);
    const total_reported = pointsOf(roster as RawSelection);

    // Honest computed total: sum every cost line in the tree. A unit's own cost
    // and its nested enhancement's cost are distinct lines that together make up
    // the unit's army contribution, so a full walk reproduces the army total.
    let total_computed = 0;
    for (const force of forces) {
      for (const sel of childSelections(force)) {
        walk(sel, (s) => {
          const pts = pointsOf(s);
          if (pts) total_computed += pts;
        });
      }
    }

    return {
      name: asString(payload.name) ?? asString(roster.name) ?? "Imported roster",
      generated_by: asString(payload.generatedBy),
      faction_raw_name: factions[0] ?? null,
      detachment_raw_names,
      battle_size_raw,
      declared_limit: parseLimit(battle_size_raw),
      total_reported,
      total_computed,
      units,
      multi_force: factions.length > 1,
    };
  },
};

File: crates/wh40kdc/src/import/listforge.rs
//! ListForge adapter: lower a decoded ListForge "share JSON" payload (a
//! BattleScribe-derived roster tree) to a [`ParsedRoster`].
//!
//! The walk reads an ALLOWLIST of fields only — `name`, `number`, `type`,
//! `categories[].name`, `group`, and `costs` point values — and never touches
//! `rules[].description` or ability `profiles[].characteristics[].$text`, which
//! carry reproduced rules text. This keeps the importer's output free of
//! copyrighted prose **by construction**; the `import_ip_safety` integration
//! test is the regression guard.
//!
//! Selection-tree shape (recursive `selections`):
//! - Configuration nodes (`type: "upgrade"`) named "Detachment" / "Battle Size"
//!   carry the chosen value as their first child selection.
//! - Unit nodes (`type: "model" | "unit"`) carry role categories, a points
//!   cost, and — nested anywhere beneath them — their wargear (weapon-category
//!   selections), enhancement (a selection whose `group` starts "Enhancements"),
//!   the "Warlord" marker, and model sub-selections.
//! - Every unit carries a `"Faction: <Name>"` category.
//!
//! Rust mirror of `tools/src/import/listforge.ts`.

use serde_json::Value;

use super::adapter::{FormatAdapter, ParseError};
use super::types::{ParsedRoster, ParsedUnit, ParsedWargear, RosterFormat};

const PTS_COST_NAME: &str = "pts";
const ENHANCEMENT_GROUP_PREFIX: &str = "Enhancements";
const WEAPON_CATEGORY_SUFFIX: &str = " Weapon"; // "Ranged Weapon", "Melee Weapon", …
const CHARACTER_CATEGORIES: [&str; 2] = ["Character", "Epic Hero"];
const NEWRECRUIT_XMLNS: &str = "http://www.battlescribe.net/schema/rosterSchema";
const NEWRECRUIT_HOST_PREFIX: &str = "https://newrecruit";

// --- Allowlisted field accessors (the IP-safety boundary). ------------------
// Only these fields of a selection are ever read. Adding an accessor here is the
// only way to widen what the importer touches, so the allowlist stays auditable.

fn as_array(value: &Value) -> &[Value] {
    value.as_array().map(Vec::as_slice).unwrap_or(&[])
}

fn as_string(value: &Value) -> Option<&str> {
    value.as_str()
}

fn selection_name(sel: &Value) -> &str {
    as_string(&sel["name"]).unwrap_or("")
}

fn selection_type(sel: &Value) -> &str {
    as_string(&sel["type"]).unwrap_or("")
}

/// A selection's multiplicity (`number`), defaulting to 1.
fn selection_count(sel: &Value) -> u64 {
    match sel["number"].as_u64() {
        Some(n) if n > 0 => n,
        _ => 1,
    }
}

/// Point value from a selection's cost block, or `None` when absent.
fn points_of(sel: &Value) -> Option<u64> {
    for cost in as_array(&sel["costs"]) {
        if as_string(&cost["name"]) == Some(PTS_COST_NAME) {
            // ListForge encodes points as a (sometimes fractional) number; the
            // 40kdc model uses whole points, so truncate toward zero.
            if let Some(v) = cost["value"].as_f64() {
                return Some(v as u64);
            }
        }
    }
    None
}

fn category_names(sel: &Value) -> Vec<&str> {
    as_array(&sel["categories"])
        .iter()
        .filter_map(|c| as_string(&c["name"]))
        .collect()
}

fn child_selections(sel: &Value) -> &[Value] {
    as_array(&sel["selections"])
}

/// Depth-first visit of a selection and everything beneath it.
fn walk(sel: &Value, visit: &mut impl FnMut(&Value)) {
    visit(sel);
    for child in child_selections(sel) {
        walk(child, visit);
    }
}

fn is_unit_selection(sel: &Value) -> bool {
    matches!(selection_type(sel), "model" | "unit")
}

fn is_character(sel: &Value) -> bool {
    category_names(sel)
        .iter()
        .any(|n| CHARACTER_CATEGORIES.contains(n))
}

fn is_weapon_selection(sel: &Value) -> bool {
    category_names(sel)
        .iter()
        .any(|n| n.ends_with(WEAPON_CATEGORY_SUFFIX))
}

fn is_enhancement_selection(sel: &Value) -> bool {
    as_string(&sel["group"]).is_some_and(|g| g.starts_with(ENHANCEMENT_GROUP_PREFIX))
}

/// Sum the model count of a unit from its nested model selections.
fn model_count(unit: &Value) -> u64 {
    let mut total = 0;
    walk(unit, &mut |s| {
        if selection_type(s) == "model" {
            total += selection_count(s);
        }
    });
    if total > 0 {
        total
    } else {
        selection_count(unit)
    }
}

/// Build a parsed unit from a top-level unit selection.
fn parse_unit(unit: &Value) -> ParsedUnit {
    let mut wargear: Vec<ParsedWargear> = Vec::new();
    let mut enhancement_raw_name: Option<String> = None;
    let mut enhancement_points: Option<u64> = None;
    let mut is_warlord = false;

    for node in child_selections(unit) {
        walk(node, &mut |s| {
            if is_enhancement_selection(s) {
                if enhancement_raw_name.is_none() {
                    enhancement_raw_name = Some(selection_name(s).to_string());
                    enhancement_points = points_of(s);
                }
                return;
            }
            if selection_name(s) == "Warlord" {
                is_warlord = true;
                return;
            }
            if is_weapon_selection(s) {
                wargear.push(ParsedWargear {
                    raw_name: selection_name(s).to_string(),
                    count: selection_count(s),
                });
            }
        });
    }

    ParsedUnit {
        raw_name: selection_name(unit).to_string(),
        is_character: is_character(unit),
        model_count: model_count(unit),
        points: points_of(unit),
        is_warlord,
        enhancement_raw_name,
        enhancement_points,
        wargear,
    }
}

/// Value carried as the first child of a named configuration selection.
fn config_value(selections: &[Value], config_name: &str) -> Option<String> {
    let node = selections
        .iter()
        .find(|s| selection_name(s) == config_name)?;
    let child = child_selections(node).first()?;
    Some(selection_name(child).to_string())
}

/// Every value under a named config, across repeated blocks and multiple
/// children, in source order. Used for multi-detachment 11e lists.
fn config_values(selections: &[Value], config_name: &str) -> Vec<String> {
    let mut out = Vec::new();
    for node in selections {
        if selection_name(node) != config_name {
            continue;
        }
        for child in child_selections(node) {
            let name = selection_name(child);
            if !name.is_empty() {
                out.push(name.to_string());
            }
        }
    }
    out
}

/// Parse the points ceiling out of a battle-size label like
/// "2. Strike Force (2000 Point limit)".
fn parse_limit(label: Option<&str>) -> Option<u64> {
    let label = label?;
    // Find a digit run immediately followed (allowing spaces) by "Point".
    let bytes = label.as_bytes();
    let lower = label.to_ascii_lowercase();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i].is_ascii_digit() {
            let start = i;
            while i < bytes.len() && (bytes[i].is_ascii_digit() || bytes[i] == b',') {
                i += 1;
            }
            // Skip whitespace then check for "point".
            let mut j = i;
            while j < bytes.len() && bytes[j] == b' ' {
                j += 1;
            }
            if lower[j..].starts_with("point") {
                let digits: String = label[start..i].chars().filter(|c| *c != ',').collect();
                return digits.parse().ok();
            }
        } else {
            i += 1;
        }
    }
    None
}

/// All distinct `"Faction: X"` category names found anywhere in the forces.
fn collect_factions(forces: &[Value]) -> Vec<String> {
    let mut seen: Vec<String> = Vec::new();
    for force in forces {
        for sel in child_selections(force) {
            walk(sel, &mut |s| {
                for name in category_names(s) {
                    if let Some(rest) = name.strip_prefix("Faction:") {
                        let value = rest.trim().to_string();
                        if !value.is_empty() && !seen.contains(&value) {
                            seen.push(value);
                        }
                    }
                }
            });
        }
    }
    seen
}

/// The roster object (with a `forces` array), if the payload carries one.
fn roster_of(decoded: &Value) -> Option<&Value> {
    let roster = decoded.get("roster")?;
    if !roster.is_object() {
        return None;
    }
    if !roster.get("forces").map(Value::is_array).unwrap_or(false) {
        return None;
    }
    Some(roster)
}

/// Detect a NewRecruit-flavoured BattleScribe payload. ListForge's matcher
/// excludes these so the greedy first-match dispatcher routes them to the
/// NewRecruit adapter without falling through to here.
fn has_newrecruit_signature(decoded: &Value, roster: &Value) -> bool {
    if as_string(&roster["xmlns"]) == Some(NEWRECRUIT_XMLNS) {
        return true;
    }
    let gen_by = as_string(&decoded["generatedBy"]).or_else(|| as_string(&roster["generatedBy"]));
    gen_by.is_some_and(|g| g.to_ascii_lowercase().starts_with(NEWRECRUIT_HOST_PREFIX))
}

/// The ListForge "share JSON" adapter — the first concrete
/// [`FormatAdapter`](super::FormatAdapter).
pub struct ListForgeAdapter;

impl FormatAdapter for ListForgeAdapter {
    fn format(&self) -> RosterFormat {
        RosterFormat::Listforge
    }

    fn detect(&self, decoded: &Value) -> bool {
        let Some(roster) = roster_of(decoded) else {
            return false;
        };
        !has_newrecruit_signature(decoded, roster)
    }

    fn parse(&self, decoded: &Value) -> Result<ParsedRoster, ParseError> {
        let roster = roster_of(decoded)
            .ok_or_else(|| ParseError("listforge: payload has no roster.forces array".into()))?;
        let forces = as_array(&roster["forces"]);

        // Configuration lives among each force's top-level selections.
        let mut detachment_raw_names: Vec<String> = Vec::new();
        let mut battle_size_raw: Option<String> = None;
        let mut units: Vec<ParsedUnit> = Vec::new();
        for force in forces {
            let top = child_selections(force);
            detachment_raw_names.extend(config_values(top, "Detachment"));
            if battle_size_raw.is_none() {
                battle_size_raw = config_value(top, "Battle Size");
            }
            for sel in top {
                if is_unit_selection(sel) {
                    units.push(parse_unit(sel));
                }
            }
        }

        let factions = collect_factions(forces);
        let total_reported = points_of(roster);

        // Honest computed total: sum every cost line in the tree. A unit's own
        // cost and its nested enhancement's cost are distinct lines that together
        // make up the unit's army contribution, so a full walk reproduces the
        // army total.
        let mut total_computed = 0;
        for force in forces {
            for sel in child_selections(force) {
                walk(sel, &mut |s| {
                    if let Some(pts) = points_of(s) {
                        total_computed += pts;
                    }
                });
            }
        }

        let name = as_string(&decoded["name"])
            .or_else(|| as_string(&roster["name"]))
            .unwrap_or("Imported roster")
            .to_string();

        Ok(ParsedRoster {
            name,
            generated_by: as_string(&decoded["generatedBy"]).map(str::to_string),
            faction_raw_name: factions.first().cloned(),
            detachment_raw_names,
            battle_size_raw: battle_size_raw.clone(),
            declared_limit: parse_limit(battle_size_raw.as_deref()),
            total_reported,
            total_computed,
            units,
            multi_force: factions.len() > 1,
        })
    }
}
